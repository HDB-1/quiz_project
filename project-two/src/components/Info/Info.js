import React from 'react'
import Progress from './Progress/Progress'
import User from './User/User'
import './Info.css'

export default function Info(props) {
    return (
        <div className='info'>
            <User currentUser={props.currentUser}/>
            <h3 id="quizName">{(props.title)}</h3>
            <Progress question={{current: props.question.current, total: props.question.total}}/>
        </div>
    )
}
