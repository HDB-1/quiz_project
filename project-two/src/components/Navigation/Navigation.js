import React from 'react'

export default function Navigation(props) {
    return (
        <div>
            <button id="nextBtn" onClick={props.nextQuestion}></button>
            <button id="previousBtn" onClick={props.previousQuestion}></button>
        </div>
    )
}
