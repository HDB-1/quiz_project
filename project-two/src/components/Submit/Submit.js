import React from 'react'
import './Submit.css'
export default function Submit(props) {
    return (
        <div className='submit'>
            <button id='skipBtn' onClick={props.skip}>Skip</button>
            <button id='submitBtn' onClick={() => props.submit(props.answerSelected)}>Submit</button>
        </div>
    )
}
