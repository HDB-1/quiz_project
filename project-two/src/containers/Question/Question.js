import React, { Component } from 'react'
import Answers from '../../components/Answers/Answers'
import Submit from '../../components/Submit/Submit'
export class Question extends Component {

    render() {
        return (
            <div>
                <h4 id="questionName">{this.props.questionInfo.question}</h4>
                <Answers answers={["1","2","3","4"]} 
                correct_answer={this.props.questionInfo.correct_answer}
                incorrect_answers={this.props.questionInfo.incorrect_answers}
                />
                <Submit skip={this.props.skip} submit={this.props.submit}/>
            </div>
        )
    }
}
export default Question