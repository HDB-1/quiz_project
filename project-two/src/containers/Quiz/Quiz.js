import React, { Component } from "react";
import Info from "../../components/Info/Info";
import Navigation from "../../components/Navigation/Navigation";
import Question from "../../containers/Question/Question";
import { Route, Switch, Link } from "react-router-dom";
import Results from "../../containers/Results/Results";

class Quiz extends Component {
  
  state = {
    currentQuestionIndex: 0,
    questions: [],
    correctAnswers: [],
    userAnswers: [],
    currentPlayerIndex: 0 
  };
  componentDidMount() {
    for(let i = 0; i < this.props.quizInfo.numOfPlayers; i++){
      this.state.userAnswers.push([])
    }
    this.APIRequest(this.props.quizInfo);
  }

  submitQuestion = answerSelected => {
    let answerFound = false;

    // make an answer
    let newAnswer = {
      content: answerSelected,
      index: this.state.currentQuestionIndex
    };
    let newAnswerArray = this.state.userAnswers[this.state.currentPlayerIndex]; // new answers array
    //check if index of new answer is in answer array
    //if not add it to the array
    //if so update array value
    if (newAnswerArray.length > 0) {
      // if there are any answers
      newAnswerArray.map(answer => {
        if (answer.index === newAnswer.index) {
          answer.content = newAnswer.content;
          answerFound = true;
        }
        return answer;
      });
    }
    if (!answerFound) {
      newAnswerArray.push(newAnswer);
    }
    let allUserAnswers = this.state.userAnswers;

    allUserAnswers[this.state.currentPlayerIndex] = newAnswerArray
    this.setState({ userAnswers: allUserAnswers });
    if(this.state.currentPlayerIndex === Number(this.props.quizInfo.numOfPlayers) - 1){
      this.nextQuestion();
      this.setState({currentPlayerIndex : 0})
    }
    else this.setState({currentPlayerIndex: this.state.currentPlayerIndex +1})
  };

  nextQuestion = () => {
    if (this.state.currentQuestionIndex < this.state.questions.length - 1) {
      this.setState({
        currentQuestionIndex: this.state.currentQuestionIndex + 1
      });
    }
  };
  previousQuestion = () => {
    if (this.state.currentQuestionIndex > 0) {
      this.setState({
        currentQuestionIndex: this.state.currentQuestionIndex - 1
      });
    }
  };

  APIRequest = quizInfo => {
    const baseUrl = "https://opentdb.com/api.php?type=multiple&";
    let url =
      baseUrl +
      `category=${quizInfo.category}&` +
      `amount=${quizInfo.numOfQuestions}&` +
      `difficulty=${quizInfo.difficulty}&` + 
      'encode=base64'
      ;
    // console.log(url);
    // console.log(fetch(url));
    fetch(url)
      .then(res => res.json())
      .then(json => {
        let correctAnswers = json.results.map(question => {
          return question.correct_answer;
        });
        this.setState({
          questions: json.results,
          correctAnswers: correctAnswers
        });
      })
      .catch(error => {
        //handle error
      });
  };

  checkIfUsersHaveAnsweredAllQuestions = () => {
    let counter = 0;
    for(let answerArray of this.state.userAnswers){
      if(answerArray.length === this.state.correctAnswers.length){
        counter ++;
      }
    }
    if(counter === Number(this.props.quizInfo.numOfPlayers)) {
      return true;
    }
    else return false;
  }

  render() {
    return (
      <Switch>
        <Route exact path="/quiz">
          <div className="quizPage">
            {this.state.questions.length > 0 ? (
              <div>
                <Info
                  title={this.state.questions[0].category}
                  currentUser={this.state.currentPlayerIndex}
                  question={{
                    current: this.state.currentQuestionIndex,
                    total: this.props.quizInfo.numOfQuestions
                  }}

                />
                <Question
                  skip={this.skipQuestion}
                  submit={this.submitQuestion}
                  questionInfo={
                    this.state.questions[this.state.currentQuestionIndex]
                  }
                />
                <Navigation
                  next={this.nextQuestion}
                  previous={this.previousQuestion}
                />
              </div>
            ) : (
              <p>Loading...</p>
            )}
            {this.checkIfUsersHaveAnsweredAllQuestions()  && <Link to="/quiz/results">
              You have answered every question! Click here to go to results
              </Link>}
          </div>
        </Route>
        <Route path="/quiz/results">
          <Results userAnswers={this.state.userAnswers} correctAnswers={this.state.correctAnswers}/>
        </Route>
      </Switch>
    );
  }
}

export default Quiz;
