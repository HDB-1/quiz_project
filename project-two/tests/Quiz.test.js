import React from "react";
import { shallow, mount } from "enzyme";
import Quiz from "../src/containers/Quiz/Quiz";
import { APIRequest } from "../src/containers/Quiz/Quiz";
import Info from "../src/components/Info/Info";
import Navigation from "../src/components/Navigation/Navigation";
import Question from "../src/containers/Question/Question";
import Submit from "../src/components/Submit/Submit";

//dummy data
const quizSetup = { difficulty: "easy", category: "9", numOfQuestions: "5" };
describe("Shallow Quiz", () => {
  let wrapper;

  let questionInfo = [
    {
      category: "General Knowledge",
      difficulty: "easy",
      question: "What is the name of NASA&rsquo;s most famous space telescope?",
      correct_answer: "Hubble Space Telescope",
      incorrect_answers: ["Big Eye", "Death Star", "Millenium Falcon"]
    },
    {
      category: "General Knowledge",
      difficulty: "easy",
      question:
        "The likeness of which president is featured on the rare $2 bill of USA currency?",
      correct_answer: "Thomas Jefferson",
      incorrect_answers: [
        "Martin Van Buren",
        "Ulysses Grant",
        "John Quincy Adams"
      ]
    }
  ];

  const quizInfoTest = {
    difficulty: "easy",
    numOfQuestions: "5",
    category: "9",
    numOfPlayers: "1",
    question: "this is a question"
  };

  beforeEach(() => {
    wrapper = shallow(<Quiz quizInfo={quizInfoTest} />);
    wrapper.setProps({ questions: questionInfo });
    wrapper.setProps({ quizInfo: quizInfoTest });

    // this automatically tests for rendering without crashing.
  });

  it("Should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should render Info, Navigation and Question component", () => {
    expect(wrapper.find(Info).length).toEqual(1);
    expect(wrapper.find(Navigation).length).toEqual(1);
    expect(wrapper.find(Question).length).toEqual(1);
  });
});

//test data
const setup = {
  difficulty: "easy",
  numOfQuestions: "5",
  category: "9",
  numOfPlayers: "1"
};

describe("testing API", () => {
  beforeEach(() => {
    fetch.resetMocks();
  });
  //API Request test will fail needs to find way to export function from class
  // it("calls API and returns data to me", () => {
  //   //Mock fetch response data
  //   fetch.mockResponseOnce(JSON.stringify({ data: "12345" }));

  //   //assert on the response
  //   APIRequest(setup).then(res => {
  //     expect(res.data).toEqual("12345");
  //   });
  //   //assert on the time called and arguments given to fetch
  //   expect(fetch.mock.calls.length).toEqual(1); // fetch.mock.calls = nested array [[]] which contains the URL.
  //   expect(fetch.mock.calls[0][0]).toEqual(
  //     "https://opentdb.com/api.php?type=multiple&category=9&amount=5&difficulty=easy"
  //   );
  // });
});

describe("Mounted quiz", () => {
  let wrapper;
  beforeEach(() => (wrapper = mount(<Quiz quizInfo={quizSetup} />)));

  it("calls skip function on skip button click", () => {
    const spy = jest.spyOn(wrapper.instance(), "skipQuestion");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper
      .find(Question)
      .find(Submit)
      .find("#skipBtn")
      .simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });
  it("calls submit function on submit button click", () => {
    const spy = jest.spyOn(wrapper.instance(), "submitQuestion");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper
      .find(Question)
      .find(Submit)
      .find("#submitBtn")
      .simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
  });
});
