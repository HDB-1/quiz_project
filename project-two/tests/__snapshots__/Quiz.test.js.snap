// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow Quiz Should render correctly 1`] = `
<div>
  <div>
    <Info
      question={
        Object {
          "current": 0,
          "total": "5",
        }
      }
      title="General Knowledge"
    />
    <Question
      questionInfo={
        Object {
          "category": "General Knowledge",
          "correct_answer": "Hubble Space Telescope",
          "difficulty": "easy",
          "incorrect_answers": Array [
            "Big Eye",
            "Death Star",
            "Millenium Falcon",
          ],
          "question": "What is the name of NASA&rsquo;s most famous space telescope?",
        }
      }
      skip={[Function]}
      submit={[Function]}
    />
    <Navigation
      next={[Function]}
      previous={[Function]}
    />
     
  </div>
</div>
`;
